File: fluentcore/src/main/java/com/mcreater/fxfluent/xaml/ResourceDict.java
package com.mcreater.fxfluent.xaml;

import com.mcreater.fxfluent.brush.AbstractColorBrush;
import com.mcreater.fxfluent.brush.SolidColorBrush;
import com.mcreater.fxfluent.xaml.tags.*;
import javafx.scene.paint.Color;

import java.util.Vector;

public class ResourceDict extends Vector<AbstractContentTag<?>> {
    public final String name;
    private ResourceDict(String name) {
        this.name = name;
    }
    public static ResourceDict createEmpty(String name) {
        return new ResourceDict(name);
    }

    public <T extends AbstractContentTag<?>> T foundTag(String key, Class<T> clazz) {
        for (AbstractContentTag<?> abstractContentTag : this) {
            if (abstractContentTag.getKey().equals(key)) {
                return abstractContentTag instanceof StaticResourceRedirectContentTag ?
                        foundTag(((StaticResourceRedirectContentTag) abstractContentTag).toObject(), clazz) :  // Redirect tags
                        (T) abstractContentTag;
            }
        }
        return this == XAMLManager.getGlobal() ? null : XAMLManager.getGlobal().foundTag(key, clazz);
    }

    public AbstractColorBrush findColorBrush(String key) {
        try {
            return foundTag(key, SolidColorBrushContentTag.class).toObject();
        }
        catch (Exception ignored) {}

        try {
            return foundTag(key, LinearGradientBrushContentTag.class).toObject();
        }
        catch (Exception ignored) {}

        try {
            return new SolidColorBrush(foundTag(key, ColorContentTag.class).toObject());
        }
        catch (Exception e) {
            return new SolidColorBrush(Color.TRANSPARENT);
        }
    }
}

File: fluentcore/src/main/java/com/mcreater/fxfluent/xaml/ResourceDict.kt
package com.mcreater.fxfluent.xaml

import com.mcreater.fxfluent.brush.AbstractColorBrush
import com.mcreater.fxfluent.brush.SolidColorBrush
import com.mcreater.fxfluent.xaml.tags.*
import javafx.scene.paint.Color
import java.util.*

class ResourceDict(name: String?): Vector<AbstractContentTag<out Any>>() {
    companion object {
        @JvmStatic
        fun createEmpty(name: String?): ResourceDict {
            return ResourceDict(name)
        }
    }
    var name: String? = null
    init {
        this.name = name
    }

    fun <T : AbstractContentTag<*>?> foundTag(key: String, clazz: Class<T>?): T? {
        for (abstractContentTag in this) {
            if (abstractContentTag.getKey() == key) {
                return if (abstractContentTag is StaticResourceRedirectContentTag) foundTag(
                    abstractContentTag.toObject(),
                    clazz
                ) else  // Redirect tags
                    abstractContentTag as T
            }
        }
        return if (this === XamlManager.getGlobal()) null else XamlManager.getGlobal().foundTag(key, clazz)
    }

    fun findColorBrush(key: String): AbstractColorBrush {
        try {
            return foundTag(key, SolidColorBrushContentTag::class.java)!!.toObject()
        } catch (ignored: Exception) {
        }
        try {
            return foundTag(key, LinearGradientBrushContentTag::class.java)!!.toObject()
        } catch (ignored: Exception) {
        }
        return try {
            SolidColorBrush(foundTag(key, ColorContentTag::class.java)!!.toObject())
        } catch (e: Exception) {
            SolidColorBrush(Color.TRANSPARENT)
        }
    }
}